rust/rbac-registration/src/registration/cardano/mod.rs

```diff
@@ -413,44 +423,46 @@ fn revocations_list(
     revocations
 }
 
-/// Update the role data in the registration chain.
+/// Update the role data related fields in the registration chain.
 fn update_role_data(
-    inner: &mut RegistrationChainInner, role_set: HashMap<RoleNumber, RoleData>,
-    point_tx_idx: &PointTxnIdx,
+    role_data_history: &mut HashMap<RoleNumber, Vec<PointData<RoleData>>>,
+    role_data_record: &mut HashMap<RoleNumber, RoleDataRecord>,
+    role_set: HashMap<RoleNumber, RoleData>, point_tx_idx: &PointTxnIdx,
 ) {
-    for (number, mut data) in role_set {
-        // Update the record of role data first
-        inner
-            .all_role_data
+    for (number, data) in role_set {
+        // Update role data history, put the whole role data
+        role_data_history
             .entry(number)
             .or_default()
             .push(PointData::new(point_tx_idx.clone(), data.clone()));
 
-        // If there is new role singing key, use it, else use the old one
-        if data.signing_key().is_none() {
-            let signing_key = inner
-                .role_data
-                .get(&number)
-                .and_then(|pd| pd.data().signing_key())
-                .cloned();
-            data.set_signing_key(signing_key);
-        }
-
-        // If there is new role encryption key, use it, else use the old one
-        if data.encryption_key().is_none() {
-            let encryption_key = inner
-                .role_data
-                .get(&number)
-                .and_then(|pd| pd.data().encryption_key())
-                .cloned();
-            data.set_encryption_key(encryption_key);
+        role_data_record
+            .entry(number)
+            .or_insert(RoleDataRecord::new());
+
+        // Update role data record
+        if let Some(record) = role_data_record.get_mut(&number) {
```

Comment:
I'm not sure, but maybe it would be better to move this block to `role_data_record.entry(number).and_modify(...)` closure?.. Currently this condition is slightly confusing. because the record is always present because we inserting it above.